pkg/clustermesh/kvstoremesh/remote_cluster.go

```diff
@@ -130,8 +133,74 @@ func (rc *remoteCluster) Stop() {
 }
 
 func (rc *remoteCluster) Remove() {
-	// Cluster specific keys are not explicitly removed, but they will be
-	// disappear once the associated lease expires.
+	const retries = 5
+
+	fn, done := rc.drainJobFn(retries)
+	rc.jobGroup.Add(job.OneShot(
+		fmt.Sprintf("drain-cluster-%s", rc.name), fn,
+		job.WithRetry(retries, &job.ExponentialBackoff{Min: 2 * time.Second, Max: 60 * time.Second}),
+	))
+
+	<-done
+}
+
+// drainJobFn returns a job function responsible for draining the cached data
+// from the local kvstore.
+func (rc *remoteCluster) drainJobFn(retries uint) (job.OneShotFunc, <-chan struct{}) {
+	var retry uint
+	done := make(chan struct{})
+	keys := []string{
+		path.Join(kvstore.ClusterConfigPrefix, rc.name),
+	}
+	prefixes := []string{
+		path.Join(kvstore.SyncedPrefix, rc.name),
+		path.Join(kvstore.StateToCachePrefix(nodeStore.NodeStorePrefix), rc.name),
+		path.Join(kvstore.StateToCachePrefix(serviceStore.ServiceStorePrefix), rc.name),
+		path.Join(kvstore.StateToCachePrefix(identityCache.IdentitiesPath), rc.name),
+		path.Join(kvstore.StateToCachePrefix(ipcache.IPIdentitiesPath), rc.name),
+	}
+
+	rc.logger.Info("Remote cluster disconnected: draining cached data")
+	return func(ctx context.Context, _ cell.Health) (err error) {
+		defer func() {
+			switch {
+			case err == nil:
+				rc.logger.Info("Successfully removed all cached data from kvstore")
+				close(done)
+			case ctx.Err() != nil:
+				// We are shutting down
+				err = nil
+				close(done)
+			case retry == retries:
+				rc.logger.WithError(err).Error(
+					"Failed to remove cached data from kvstore, despite retries. Reconnecting to the " +
+						"same cluster without first restarting KVStoreMesh may lead to inconsistencies")
+				close(done)
+			default:
+				rc.logger.WithError(err).Warning("Failed to remove cached data from kvstore, retrying")
+				retry++
+			}
+
+		}()
+
+		for _, key := range keys {
+			if err = rc.localBackend.Delete(ctx, key); err != nil {
+				return err
+			}
+
+			keys = keys[1:]
+		}
+
+		for _, prefix := range prefixes {
+			if err = rc.localBackend.DeletePrefix(ctx, prefix+"/"); err != nil {
+				return err
+			}
+
+			prefixes = prefixes[1:]
```

Comment:
nit: Is there any reason why we are overriding prefixes/keys?
I know it's technically correct, but a bit strange to read.

Reply:
The idea is to avoid attempting to delete again successful prefixes/keys in case an error has occurred at a later stage. That's only a minor optimization, I can definitely drop it if you prefer.